crates/bitwarden-crypto/src/opaque_ke/types.rs

```diff
@@ -0,0 +1,168 @@
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+pub enum Ksf {
+    #[serde(rename = "argon2id")]
+    Argon2id(u32, u32, u32),
+}
+
+#[derive(Serialize, Deserialize, Default, Clone)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct Argon2Id {
+    /// Memory cost in kibibytes
+    pub memory_kib: u32,
+    /// Time cost in number of iterations
+    pub iterations: u32,
+    /// Number of threads that may be used
+    pub parallelism: u32,
+}
+
+impl opaque_ke::ksf::Ksf for Argon2Id {
+    fn hash<L: ArrayLength<u8>>(
+        &self,
+        input: GenericArray<u8, L>,
+    ) -> Result<GenericArray<u8, L>, InternalError> {
+        let mut output = GenericArray::default();
+        argon2::Argon2::new(
+            argon2::Algorithm::Argon2id,
+            argon2::Version::V0x13,
+            Params::new(
+                self.memory_kib,
+                self.iterations,
+                self.parallelism,
+                Some(output.len()),
+            )
+            .map_err(|_| InternalError::KsfError)?,
+        )
+        .hash_password_into(&input, &[0; argon2::RECOMMENDED_SALT_LEN], &mut output)
+        .map_err(|_| InternalError::KsfError)?;
+        Ok(output)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct RistrettoTripleDhArgonSuite {
+    pub ksf_params: Argon2Id,
+}
+
+impl CipherSuite for RistrettoTripleDhArgonSuite {
+    type OprfCs = opaque_ke::Ristretto255;
+    type KeGroup = opaque_ke::Ristretto255;
+    type KeyExchange = opaque_ke::key_exchange::tripledh::TripleDh;
+    type Ksf = Argon2Id;
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct ClientRegistrationStartResult {
+    /// The client side, serialized state of the registration process.
+    /// MUST NOT BE SHARED WITH THE SERVER and must be cleared from memory after registration
+    /// finish.
+    pub state: ByteBuf,
+    /// The serialized message to be sent to the server.
+    pub registration_request: ByteBuf,
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct ClientRegistrationFinishResult {
+    /// The serialized message to be sent to the server.
+    pub registration_upload: ByteBuf,
+    /// The public key of the server.
+    pub server_public_key: ByteBuf,
+    /// The client-side only export key result from the AKE.
+    pub export_key: ByteBuf,
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct ClientLoginStartResult {
+    /// The client side, serialized state of the login process.
+    /// MUST NOT BE SHARED WITH THE SERVER and must be cleared from memory after registration
+    /// finish.
+    pub state: ByteBuf,
```

Comment:
Sounds like this state should be encapsulated by the SDK and not leave it's boundary?